File: tests/java_static/src/test/java/myairtable/TestClientHardening.java
package myairtable;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * J-port of Kotlin myairtable-jz87 (TestClientHardening.kt) — close() ownership, retry/backoff,
 * Retry-After cap / HTTP-date handling, transport-error wrapping, and credential validation.
 *
 * <p>Adaptation notes vs the Kotlin twin:
 *
 * <ul>
 *   <li>Ktor's {@code MockEngine} becomes {@link FakeTransport}; {@code runTest}'s virtual clock
 *       has no JDK equivalent, so waits are real ({@code Thread.sleep}) with tiny delays. Exact
 *       virtual-time equalities become discriminating real-clock bounds: a lower bound proves the
 *       wait happened; regressions to the wrong path are caught by configuring the alternative
 *       (jittered backoff) to be multiple seconds and asserting the elapsed time stayed far below.
 *   <li>{@code closeClosesOwnedEngine}: Kotlin asserts the closed-engine failure is NOT {@code
 *       Network}; the JDK HttpClient throws {@code IOException("closed")} after {@code close()},
 *       which the runtime wraps as {@link AirtableException.Network}. The closed-engine failure is
 *       instead distinguished from a live connection attempt by its cause message ({@code
 *       "closed"}, not a connection-refused error).
 * </ul>
 */
class TestClientHardening {

  private static final String OK_BODY = "{\"records\": []}";

  private static AirtableClient client(
      FakeTransport transport,
      int maxRetries,
      double baseRetryDelaySeconds,
      double maxRetryDelaySeconds) {
    return new AirtableClient(
        "appX",
        "key",
        "https://api.airtable.com/v0",
        new CacheStore(),
        maxRetries,
        baseRetryDelaySeconds,
        maxRetryDelaySeconds,
        30.0,
        transport);
  }

  /** Transport that answers 429 (with optional Retry-After) {@code failures} times, then 200. */
  private static FakeTransport flakyTransport(int failures, String retryAfter) {
    return new FakeTransport(
        (request, callIndex) -> {
          if (callIndex < failures) {
            Map<String, List<String>> headers =
                retryAfter != null ? Map.of("Retry-After", List.of(retryAfter)) : Map.of();
            return new FakeTransport.Canned(
                429,
                "{\"error\": {\"type\": \"RATE_LIMITED\", \"message\": \"slow down\"}}",
                headers);
          }
          return FakeTransport.Canned.ok(OK_BODY);
        });
  }

  private static long elapsedMillis(long startNanos) {
    return (System.nanoTime() - startNanos) / 1_000_000;
  }

  // ---- Closeable ownership ----

  @Test
  void closeDoesNotCloseInjectedClient() {
    FakeTransport injected = new FakeTransport((request, i) -> FakeTransport.Canned.ok(OK_BODY));
    AirtableClient client = client(injected, 3, 0.5, 30.0);
    client.listRecords("tbl1", new AirtableQuery());
    client.close();
    assertFalse(injected.isClosed(), "injected client must survive AirtableClient.close()");
    // The injected client is still fully usable after close().
    client.listRecords("tbl2", new AirtableQuery());
    assertEquals(2, injected.callCount());
    // The caller closes the injected client themselves — no exception.
    injected.close();
    assertTrue(injected.isClosed());
  }

  @Test
  void closeClosesOwnedEngine() {
    // No injected client: the AirtableClient owns a real JDK HttpClient. apiBase points
    // at a dead local port so a regression (engine left open) fails fast with a
    // connection error instead of hitting the real API — a *closed* engine fails before
    // any connection attempt, with IOException("closed").
    AirtableClient client =
        new AirtableClient(
            "appX", "key", "http://127.0.0.1:9", new CacheStore(), 3, 0.01, 0.05, 5.0, null);
    client.close();
    client.close(); // idempotent
    AirtableException.Network err =
        assertThrows(
            AirtableException.Network.class,
            () -> client.listRecords("tbl1", new AirtableQuery()),
            "request on a closed owned client must fail");
    assertEquals(
        "closed",
        err.getCause().getMessage(),
        "failure must come from the closed engine, not a live connection attempt");
  }

  // ---- Credential validation ----

  @Test
  void blankApiKeyThrowsMissingCredentials() {
    assertThrows(AirtableException.MissingCredentials.class, () -> new AirtableClient("appX", ""));
  }

  @Test
  void blankBaseIdThrowsMissingCredentials() {
    assertThrows(AirtableException.MissingCredentials.class, () -> new AirtableClient("  ", "key"));
  }

  // ---- Retry behavior ----

  @Test
  void rateLimitedThenSuccessRetriesWithJitteredBackoff() {
    FakeTransport transport = flakyTransport(1, null);
    AirtableClient client = client(transport, 3, 0.1, 30.0);
    long start = System.nanoTime();
    String payload = client.listRecords("tbl1", new AirtableQuery());
    long elapsed = elapsedMillis(start);
    assertEquals(OK_BODY, payload);
    assertEquals(2, transport.callCount());
    // No Retry-After header: backoff = 0.1s * 2^0, jittered into [50ms, 100ms).
    assertTrue(
        elapsed >= 50, "jittered backoff expected to wait at least 50ms, got " + elapsed + "ms");
    assertTrue(elapsed < 5000, "jittered backoff expected well under 5s, got " + elapsed + "ms");
  }

  @Test
  void numericRetryAfterIsHonoredExactly() {
    FakeTransport transport = flakyTransport(1, "0.05");
    // base 10s: if the Retry-After were ignored, the jittered backoff would wait >= 5s.
    AirtableClient client = client(transport, 3, 10.0, 30.0);
    long start = System.nanoTime();
    String payload = client.listRecords("tbl1", new AirtableQuery());
    long elapsed = elapsedMillis(start);
    assertEquals(OK_BODY, payload);
    // Server-provided value is honored exactly (no jitter): ~50ms, far below the 5s+ backoff.
    assertTrue(elapsed >= 50, "Retry-After: 0.05 must wait at least 50ms, got " + elapsed + "ms");
    assertTrue(
        elapsed < 5000,
        "Retry-After: 0.05 must preempt the jittered backoff, got " + elapsed + "ms");
  }

  @Test
  void absurdRetryAfterIsCappedAtMaxRetryDelay() {
    FakeTransport transport = flakyTransport(1, "3600");
    AirtableClient client = client(transport, 3, 0.01, 0.05);
    long start = System.nanoTime();
    String payload = client.listRecords("tbl1", new AirtableQuery());
    long elapsed = elapsedMillis(start);
    assertEquals(OK_BODY, payload);
    assertTrue(
        elapsed >= 50,
        "capped Retry-After must still wait maxRetryDelaySeconds (50ms), got " + elapsed + "ms");
    assertTrue(
        elapsed < 5000,
        "Retry-After: 3600 must be capped to maxRetryDelaySeconds (50ms), got " + elapsed + "ms");
  }

  @Test
  void httpDateRetryAfterInThePastWaitsZero() {
    FakeTransport transport = flakyTransport(1, "Wed, 21 Oct 2015 07:28:00 GMT");
    // base 10s: a regression to the jittered-backoff path would wait >= 5s.
    AirtableClient client = client(transport, 3, 10.0, 30.0);
    long start = System.nanoTime();
    String payload = client.listRecords("tbl1", new AirtableQuery());
    long elapsed = elapsedMillis(start);
    assertEquals(OK_BODY, payload);
    assertEquals(2, transport.callCount());
    assertTrue(
        elapsed < 5000, "HTTP-date in the past must floor the wait at 0, got " + elapsed + "ms");
  }

  @Test
  void terminal429ThrowsRateLimitedWithParsedRetryAfter() {
    FakeTransport transport = flakyTransport(10, "0.01");
    AirtableClient client = client(transport, 1, 0.01, 30.0);
    AirtableException.RateLimited err =
        assertThrows(
            AirtableException.RateLimited.class,
            () -> client.listRecords("tbl1", new AirtableQuery()));
    assertEquals(0.01, err.retryAfterSeconds());
  }

  // ---- Network exception wrapping ----

  @Test
  void transportIOExceptionSurfacesAsNetwork() {
    FakeTransport transport =
        new FakeTransport(
            (request, i) -> {
              throw new IOException("connection reset");
            });
    AirtableClient client = client(transport, 3, 0.01, 30.0);
    AirtableException.Network err =
        assertThrows(
            AirtableException.Network.class, () -> client.listRecords("tbl1", new AirtableQuery()));
    assertInstanceOf(IOException.class, err.getCause());
    assertTrue(err.getMessage().contains("connection reset"));
  }

  @Test
  void networkIsCaughtByTheAirtableExceptionContract() {
    FakeTransport transport =
        new FakeTransport(
            (request, i) -> {
              throw new IOException("boom");
            });
    AirtableClient client = client(transport, 3, 0.01, 30.0);
    try {
      client.listRecords("tbl1", new AirtableQuery());
      fail("expected a throw");
    } catch (AirtableException e) {
      assertInstanceOf(AirtableException.Network.class, e);
    }
  }

  @Test
  void networkEqualsSemanticsMatchSiblings() {
    assertEquals(
        new AirtableException.Network(new IOException("a")),
        new AirtableException.Network(new IOException("b")));
    assertNotEquals(
        new AirtableException.Network(new IOException("a")),
        new AirtableException.MissingCredentials());
  }
}

File: tests/kotlin_static/src/test/kotlin/myairtable/TestClientHardening.kt
package myairtable

import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.isActive
import kotlinx.coroutines.test.currentTime
import kotlinx.coroutines.test.runTest
import java.io.IOException
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * myairtable-jz87 — client hardening: Closeable ownership, retry-loop body
 * consumption, Retry-After cap / HTTP-date / jitter, transport-error
 * wrapping, and credential validation.
 *
 * Timing assertions use runTest's virtual clock ([currentTime]): `delay()`
 * inside the client is skipped instantly but still advances virtual time, so
 * waits are asserted exactly without real sleeps.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class TestClientHardening {
    private val okBody = """{"records": []}"""

    /** MockEngine that answers 429 (with optional Retry-After) [failures] times, then 200. */
    private fun flakyEngine(
        failures: Int,
        retryAfter: String? = null,
        onCall: () -> Unit = {},
    ): MockEngine {
        var calls = 0
        return MockEngine {
            calls += 1
            onCall()
            if (calls <= failures) {
                respond(
                    content = """{"error": {"type": "RATE_LIMITED", "message": "slow down"}}""",
                    status = HttpStatusCode.TooManyRequests,
                    headers = if (retryAfter != null) headersOf(HttpHeaders.RetryAfter, retryAfter) else headersOf(),
                )
            } else {
                respond(
                    content = okBody,
                    status = HttpStatusCode.OK,
                    headers = headersOf(HttpHeaders.ContentType, "application/json"),
                )
            }
        }
    }

    // region Closeable ownership

    @Test
    fun closeDoesNotCloseInjectedClient() =
        runTest {
            val engine =
                MockEngine {
                    respond(
                        content = okBody,
                        status = HttpStatusCode.OK,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }
            val injected = HttpClient(engine)
            val client = AirtableClient(baseId = "appX", apiKey = "key", httpClient = injected)
            client.listRecords("tbl1")
            client.close()
            assertTrue(injected.isActive, "injected client must survive AirtableClient.close()")
            assertTrue(engine.isActive, "injected engine must survive AirtableClient.close()")
            // The injected client is still fully usable after close().
            client.listRecords("tbl2")
            // The caller closes the injected client themselves — no exception.
            injected.close()
        }

    @Test
    fun closeClosesOwnedEngine() =
        runTest {
            // No injected client: the AirtableClient owns a real CIO engine.
            // apiBase points at a dead local port so a regression (engine left
            // open) fails fast as AirtableException.Network instead of hitting
            // the real API — a *closed* engine fails before any connection
            // attempt, with a non-Network error.
            val client = AirtableClient(baseId = "appX", apiKey = "key", apiBase = "http://127.0.0.1:9")
            client.close()
            client.close() // idempotent
            val result = runCatching { client.listRecords("tbl1") }
            assertTrue(result.isFailure, "request on a closed owned client must fail")
            assertFalse(
                result.exceptionOrNull() is AirtableException.Network,
                "failure must come from the closed engine, not a live connection attempt",
            )
        }

    // endregion

    // region Credential validation

    @Test
    fun blankApiKeyThrowsMissingCredentials() {
        assertFailsWith<AirtableException.MissingCredentials> {
            AirtableClient(baseId = "appX", apiKey = "")
        }
    }

    @Test
    fun blankBaseIdThrowsMissingCredentials() {
        assertFailsWith<AirtableException.MissingCredentials> {
            AirtableClient(baseId = "  ", apiKey = "key")
        }
    }

    // endregion

    // region Retry behavior

    @Test
    fun rateLimitedThenSuccessRetriesWithJitteredBackoff() =
        runTest {
            var calls = 0
            val engine = flakyEngine(failures = 1) { calls += 1 }
            val client = AirtableClient(baseId = "appX", apiKey = "key", httpClient = HttpClient(engine))
            val payload = client.listRecords("tbl1")
            assertEquals(okBody, payload)
            assertEquals(2, calls)
            // No Retry-After header: backoff = 0.5s * 2^0, jittered into [0.25s, 0.5s).
            assertTrue(currentTime in 250..500, "jittered backoff expected in [250ms, 500ms), got ${currentTime}ms")
        }

    @Test
    fun numericRetryAfterIsHonoredWithBoundedJitter() =
        runTest {
            val engine = flakyEngine(failures = 1, retryAfter = "0.05")
            val client = AirtableClient(baseId = "appX", apiKey = "key", httpClient = HttpClient(engine))
            client.listRecords("tbl1")
            // Server-provided value is honored, plus a bounded decorrelation
            // jitter: wait = 0.05 + random*[0, min(0.05, RETRY_AFTER_JITTER_CAP)),
            // i.e. [50ms, 100ms). Jitter is present (spec: no path may be jitter-free)
            // but bounded so it stays a nudge, never a material delay.
            assertTrue(currentTime in 50..99, "Retry-After: 0.05 jittered into [50ms, 100ms), got ${currentTime}ms")
        }

    @Test
    fun absurdRetryAfterIsCappedAtMaxRetryDelay() =
        runTest {
            val engine = flakyEngine(failures = 1, retryAfter = "3600")
            val client =
                AirtableClient(
                    baseId = "appX",
                    apiKey = "key",
                    maxRetryDelaySeconds = 0.05,
                    httpClient = HttpClient(engine),
                )
            val payload = client.listRecords("tbl1")
            assertEquals(okBody, payload)
            assertEquals(50, currentTime, "Retry-After: 3600 must be capped to maxRetryDelaySeconds (50ms)")
        }

    @Test
    fun httpDateRetryAfterInThePastWaitsZero() =
        runTest {
            var calls = 0
            val engine = flakyEngine(failures = 1, retryAfter = "Wed, 21 Oct 2015 07:28:00 GMT") { calls += 1 }
            val client = AirtableClient(baseId = "appX", apiKey = "key", httpClient = HttpClient(engine))
            val payload = client.listRecords("tbl1")
            assertEquals(okBody, payload)
            assertEquals(2, calls)
            assertEquals(0, currentTime, "HTTP-date in the past must floor the wait at 0")
        }

    @Test
    fun terminal429ThrowsRateLimitedWithParsedRetryAfter() =
        runTest {
            val engine = flakyEngine(failures = 10, retryAfter = "0.01")
            val client =
                AirtableClient(
                    baseId = "appX",
                    apiKey = "key",
                    maxRetries = 1,
                    httpClient = HttpClient(engine),
                )
            val err =
                assertFailsWith<AirtableException.RateLimited> {
                    client.listRecords("tbl1")
                }
            assertEquals(0.01, err.retryAfterSeconds)
        }

    // endregion

    // region Network exception wrapping

    @Test
    fun transportIOExceptionSurfacesAsNetwork() =
        runTest {
            val engine = MockEngine { throw IOException("connection reset") }
            val client = AirtableClient(baseId = "appX", apiKey = "key", httpClient = HttpClient(engine))
            val err =
                assertFailsWith<AirtableException.Network> {
                    client.listRecords("tbl1")
                }
            assertTrue(err.cause is IOException)
            assertTrue(err.message!!.contains("connection reset"))
        }

    @Test
    fun networkIsCaughtByTheAirtableExceptionContract() =
        runTest {
            val engine = MockEngine { throw IOException("boom") }
            val client = AirtableClient(baseId = "appX", apiKey = "key", httpClient = HttpClient(engine))
            try {
                client.listRecords("tbl1")
                kotlin.test.fail("expected a throw")
            } catch (e: AirtableException) {
                assertTrue(e is AirtableException.Network)
            }
        }

    @Test
    fun networkEqualsSemanticsMatchSiblings() {
        assertEquals<AirtableException>(
            AirtableException.Network(IOException("a")),
            AirtableException.Network(IOException("b")),
        )
        assertFalse(AirtableException.Network(IOException("a")).equals(AirtableException.MissingCredentials()))
    }

    // endregion
}
